fix(sarif): helpUri fallback to source-rule-url

candrews opened the pull request:
See: https://github.com/semgrep/semgrep/issues/10890

candrews committed:
fix(sarif): helpUri fallback to source-rule-uri

See: https://github.com/semgrep/semgrep/issues/10890

mjambon commented:
We should explain in this code why two different fields are read from the `metadata` section and serve the same purpose. I suppose one should be deprecated in favor of the other but the structure of the `metadata` section [isn't documented](https://semgrep.dev/docs/writing-rules/rule-syntax#metadata). I asked rule authors to chime in, please hold on.

candrews commented:
> I asked rule authors to chime in, please hold on.

Do you have any news?

I'm happy to modify this MR, or someone else is of course welcome to do too (I wouldn't mind!).

LewisArdern commented on src/osemgrep/reporting/Sarif_output.ml:
```diff
@@ -143,7 +143,11 @@ let rule ~(hide_nudge : bool) (ctx : Out.format_context) (rule : Rule.t) :
     | Some _
     | None ->
-        None
+        match JSON.member "source-rule-uri" metadata with
```
This is wrong its source-rule-url

mjambon replied:
ok, so this change does what we want. We'll document the `metadata` section of Semgrep rules, eventually.

Please add a comment explaining the following since it's not documented:
```
(*
  In a Semgrep rule's metadata section, two fields may provide URLs:
  - source: populated dynamically by the Semgrep registry serving the rule, it's a URL that 
    offers information about the rule.
  - source-rule-url: optional string, a URL for the source of inspiration for the rule.

  The SARIF format supports only one URL under the field 'helpUri'. Semgrep populates it with
  metadata.source if available, metadata.source-rule-url otherwise as a fallback.
*)
```

candrews replied:
I changed it to `source-rule-url` (oops) and added the comment.

LewisArdern commented:
The Semgrep registry injects the source metadata field at runtime when you download rules from rulepacks or policies, but if people are doing local scans with local rules, the source would be blank.

As long as source metadata fields takes priority, thats fine to fall back to source-rule-url since it is useful information to know about the issue since its good information to know about the vulnerability or origin.

mjambon commented:
Sorry for the delay, @candrews. See the comments above.

I'll let you add a changelog entry as a file named `changelog.d/gh-10891.fixed` for the Semgrep release notes.

candrews committed:
fix(sarif): helpUri fallback to source-rule-url

candrews committed:
docs(sarif): comment that SARIF helpUri comes from source/source-rule-url

candrews committed:
docs(sarif): changelog for gh-10891, helpUri fallback to source-rule-url

candrews commented:
> I'll let you add a changelog entry as a file named `changelog.d/gh-10891.fixed` for the Semgrep release notes.

I added the changelog file.

Thank you!

mjambon reviewed (approved):
Thank you for your contribution to Semgrep!

mjambon commented:
To fix "[lint / pre-commit](https://github.com/semgrep/semgrep/actions/runs/14107968901/job/39552855689?pr=10891)", run `pre-commit install` and then `pre-commit run -a`.

mjambon commented:
To fix the tests, `cd cli && make regenerate-tests` should do it. I don't know what's going on with the submodule check.

One of us can take over your branch if you prefer, let us know.

candrews committed:
chore: fix lint findings

candrews commented:
> To fix the tests, `cd cli && make regenerate-tests` should do it. I don't know what's going on with the submodule check.

I'm having a  very hard time getting that done. I've tried `make dev-setup` (which appears to be a prerequisite for  `cd cli && make regenerate-tests`) a few ways, but haven't been successful.

On Fedora 42, it fails with:
```
[ERROR] The compilation of ppxlib.0.35.0 failed at "dune build -p ppxlib -j 11 @install".

#=== ERROR while compiling ppxlib.0.35.0 ======================================#
# context     2.3.0 | linux/x86_64 | ocaml.5.1.1 | https://opam.ocaml.org#251c7662afb33414976c24f2669700aae64aa4b7
# path        ~/.opam/default/.opam-switch/build/ppxlib.0.35.0
# command     ~/.opam/opam-init/hooks/sandbox.sh build dune build -p ppxlib -j 11 @install
# exit-code   1
# env-file    ~/.opam/log/ppxlib-3348831-592566.env
# output-file ~/.opam/log/ppxlib-3348831-592566.out
### output ###
# (cd _build/default && /home/candrews/.opam/default/bin/ocamlc.opt -w -9 -g -bin-annot -I astlib/.astlib.objs/byte -I /home/candrews/.opam/default/lib/ocaml-compiler-libs/common -I /home/candrews/.opam/default/lib/ocaml/compiler-libs -no-alias-deps -open Astlib__ -o astlib/.astlib.objs/byte/astlib__Ast_501.cmo -c -impl astlib/ast_501.pp.ml)
# File "astlib/ast_501.ml", lines 245-299, characters 2-44:
# Error: This variant or record definition does not match that of type
#          Parsetree.pattern_desc
#        An extra constructor, Ppat_effect, is provided in the original definition.
```

On Alpine, it fails with:
```
[ERROR] Package conflict!
  * Incompatible packages:
    - (invariant) -> ocaml-variants = 5.2.1+options -> ocaml-option-bytecode-only
    - deps-of-semgrep -> ocurl = 0.9.1
    You can temporarily relax the switch invariant with `--update-invariant'
```

> One of us can take over your branch if you prefer, let us know.

In the interest of getting this done, could you please take it across the finish line for me?

Thank you for working with me and getting this done!

mjambon commented:
> In the interest of getting this done, could you please take it across the finish line for me?

Absolutely. I feel your pain. I'll get to it on Monday.

candrews commented:
> I'll get to it on Monday.

Happy Monday! :crossed_fingers: that you're on track, thank you again!

mjambon commented:
The new PR is here: https://github.com/semgrep/semgrep/pull/10934

candrews commented:
Closing in favor of https://github.com/semgrep/semgrep/pull/10934
